### PR TITLE
structs: fix optional primitives

### PR DESCRIPTION
**Pull request by abhinav**
Previously the generated code for optional primitive fields was:

    *s.SomeField, err := ...

Which is invalid because SomeField is probably `nil` and dereferencing that is
invalid. The new logic code for this is,

    x, err := ...
    s.SomeField = &x

### CODE REVIEW

**Comment by abhinav** [comment]
@prashantv @breerly @akshayjshah

File: gen/struct_test.go [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+				{ID: 4, Value: wire.NewValueI32(3)},
+				{ID: 5, Value: wire.NewValueI64(4)},
+				{ID: 6, Value: wire.NewValueDouble(5.0)},
+				{ID: 7, Value: wire.NewValueString("foo")},
+				{ID: 8, Value: wire.NewValueBinary([]byte("bar"))},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{BoolField: boolp(true)},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
```
Comment (prashantv) [inline gen/struct_test.go]: optional: you might want a helper to reduce the line noise in the tests
```go
singleFieldStruct := func(id int, value wire.Value) wire.Value {
  return wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
    {ID: id, Value: value},
  }})
}
```

File: gen/struct_test.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+						}),
+					}),
+				},
+				{
+					ID: 3,
+					Value: wire.NewValueMap(wire.Map{
+						KeyType:   wire.TI64,
+						ValueType: wire.TDouble,
+						Size:      2,
+						Items: wire.MapItemListFromSlice([]wire.MapItem{
```
Comment (prashantv) [inline gen/struct_test.go]: does serialization of the map lead to a consistent ordering of map items? Or does it use range to iterate over the map, which could lead to different orders in the resulting wire value.

Comment (abhinav) [inline gen/struct_test.go]: Uses range. Not consistent. `ValuesAreEqual` makes unordered comparison.

File: gen/list.go [hunk 3]
```diff
@@ -123,7 +123,6 @@ func (l listGenerator) Reader(g Generator, spec *compile.ListSpec) (string, erro
 
 				<$o> := make(<$listType>, 0, <$l>.Size)
 				err := <$l>.Items.ForEach(func(<$x> <$wire>.Value) error {
-					var err error
```
Comment (prashantv) [inline gen/list.go]: Shouldn't this change have caused `structs.thrift.go` golden test to fail?

Comment (prashantv) [inline gen/list.go]: nvm, diff was suppressed, the file did change as expected.

**Comment by prashantv** [comment]
lgtm

File: gen/struct_test.go (hunk 1 above)
Comment (abhinav) [inline gen/struct_test.go]: :+1:

**Pull request merged**